lib/rex/post/postgresql/ui/console/command_dispatcher/client.rb

```diff
@@ -0,0 +1,166 @@
+            'datdba' => 'Owner ID',
+            'encoding' => 'Encoding',
+            'datlocprovider' => 'Locale Provider',
+            'datistemplate' => 'Template?',
+            'datallowconn' => 'Allow Connections?',
+            'datconnlimit' => 'Connection Limit',
+            'datfrozenxid' => 'Frozen XID',
+            'datminmxid' => 'Multixact ID',
+            'dattablespace' => 'Tablespace ID',
+            'datcollate' => 'LC_COLLATE',
+            'datctype' => 'LC_CTYPE',
+            'daticulocale' => 'ICU Locale',
+            'daticurules' => 'ICU Rules',
+            'datcollversion' => 'Collation Version',
+            'datacl' => 'Access Privileges'
+          }.freeze
+
+          #
+          # Initializes an instance of the core command set using the supplied console
+          # for interactivity.
+          #
+          # @param [Rex::Post::PostgreSQL::Ui::Console] console
+          def initialize(console)
+            super
+
+            @db_search_results = []
+          end
+
+          #
+          # List of supported commands.
+          #
+          def commands
+            cmds = {
+              'query'   => 'Run a raw SQL query',
+              'shell'   => 'Enter a raw shell where SQL queries can be executed',
+            }
+
+            reqs = {}
+
+            filter_commands(cmds, reqs)
+          end
+
+          def name
+            'PostgreSQL Client'
+          end
+
+          def help_args?(args)
+            return false unless args.instance_of?(::Array)
+
+            args.include?('-h') || args.include?('--help')
+          end
+
+          def cmd_shell_help
+            print_line 'Usage: shell'
+            print_line
+            print_line 'Go into a raw SQL shell where SQL queries can be executed.'
+            print_line 'To exit, type `exit`, `quit`, `end` or `stop`.'
+            print_line
+          end
+
+          def cmd_shell(*args)
+            cmd_shell_help && return if help_args?(args)
+
+            use_history = true
+            prompt_proc_before = ::Reline.prompt_proc
+
+            ::Reline.prompt_proc = proc { |line_buffer| line_buffer.each_with_index.map { |_line, i| i > 0 ? 'SQL *> ' : 'SQL >> ' } }
+
+            stop_words = %w[stop s exit e end quit q].freeze
+
+            finished = false
+            until finished
+              begin
+                raw_query = ::Reline.readmultiline('SQL >> ', use_history) do |multiline_input|
+                  if stop_words.include?(multiline_input.split.last)
+                    finished = true
+                    true
+                  end
+                  !multiline_input.split.last.end_with?('\\')
+                end
+              rescue ::Interrupt
+                finished = true
+              ensure
+                ::Reline.prompt_proc = prompt_proc_before
+              end
+
+              if finished
+                print_status 'Exiting Shell mode.'
+                return
+              end
```

Comment:
you have both this and an `until finished` loop I don't believe you need this but I might be missing something

Reply:
Hm, I believe this is just acting as an interactive REPL so it needs to be in a loop, i.e. READ/EVAL/PRINT/**LOOP**

Throw on a code change suggestion thought if we've misunderstood your intent though 💯

Reply:
@dwelch-r7 Ping 🏓